core/src/main/scala/dagr/core/execsystem2/Executor.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package dagr.core.execsystem2
+
+import dagr.commons.util.LazyLogging
+import dagr.core.tasksystem.{Task, UnitTask}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future}
+import scala.util.{Failure, Success}
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+
+/** Coordinates between the dependency graph ([[DependencyGraph]]) and task executor ([[TaskExecutor]]) given a (root)
+  * task to execute.
+  */
+trait Executor {
+  // start the execution by adding task that is the root of all tasks
+  def execute(rootTask: Task): Int
+
+  // the thing that execute tasks
+  protected def taskExecutor: TaskExecutor
+
+  // the thing that tracks dependencies
+  protected def dependencyGraph: DependencyGraph
+}
+
+object ExecutorImpl {
+  /** A little class to store an exception and associated task status. */
+  case class TaggedException(thr: Throwable, status: TaskStatus) extends Exception
+
+  /** An implicit to help track the failure of a future and associate it with a task status. */
+  implicit class FailWithTaskStatus[T](future: Future[T]) {
+    def failWithTaskStatus(status: TaskStatus): Future[T] = {
+      future recoverWith { case thr => Future.failed[T](TaggedException(thr=thr, status=status)) }
+    }
+  }
+}
+
+class ExecutorImpl(protected val taskExecutor: TaskExecutor) extends Executor with LazyLogging {
```

Comment:
I am using `Future`s for callbacks, and I am a bit of a novice here.  But I think the idea is sound.